src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,250 @@
+  }
+  if (GetKeyState(VK_NUMLOCK) & 0x0001) {
+    *mods = *mods | kNumLock;
+  }
+  if (GetKeyState(VK_SCROLL) & 0x0001) {
+    *mods = *mods | kScrollLock;
+  }
+  return true;
+}
+
+#elif defined(__APPLE__)
+/// Get information about Num/Caps Lock state in MacOS
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmask
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *err = NULL;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool stateCaps, stateNums;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault,
+                                    (CFDictionaryRef)mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    *err = "Unable to start service to query data";
+    return false;
+  }
+
+  if (IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType,
+                    &ioc) != kIOReturnSuccess) {
+    IOObjectRelease(ios);
+    *err = "Unable to query data";
+    return false;
+  }
+  IOObjectRelease(ios);
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateCaps) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateNums) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+  if (stateCaps) {
+    *mods = *mods | kCapsLock;
+  }
+  if (stateNums) {
+    *mods = *mods | kNumLock;
+  }
+  IOServiceClose(ioc);
+  return true;
+}
+#else
+/// A dummy function to be used by os_mod_status() if none of the above
+/// platforms were detected
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmask. Sets to 0 for dummy
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error. No error detected for dummy
+///
+/// @return true for all the cases
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *mods = 0;
+  *err = NULL;
+  return true;
+}
+#endif
+
+/// Get information about Num/Caps/Scroll Lock state
+///
+/// To be used in nvim_get_keyboard_mods_state() function.
+///
+/// @param[out]  dict  Pointer to dictionary where information about modifiers
+///                    is to be dumped.
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+bool os_mods_status(Dictionary *const dict, const char **const err)
```

Comment:
Since there already is `Dictionary` then it is fine to carry `Error*` over as well. And `Error` *does* have a buffer you could format messages into, with `api_set_error` function being already printf-like.